src/index.js

```diff
@@ -62,134 +117,128 @@ class Masq {
+    const currentUserInfo = window.sessionStorage.getItem(CURRENT_USER_INFO_STR)
+    if (currentUserInfo) {
+      await auxLoginFromUserInfo(currentUserInfo)
+      return
+    }
 
-    this._logIntoMasqErr = null
-    this._logIntoMasqDone = false
+    // if userId is is not in session storage, look for it in local storage
+    // and save in session storage
+    const localStorageCurrentUserInfo = window.localStorage.getItem(CURRENT_USER_INFO_STR)
+    if (localStorageCurrentUserInfo) {
+      await auxLoginFromUserInfo(localStorageCurrentUserInfo)
+      window.sessionStorage.setItem(CURRENT_USER_INFO_STR, localStorageCurrentUserInfo)
+    }
   }
 
-  isLoggedIn () {
-    // is logged in if the userAppDbId is known
-    // (either after register or after construction and load session info)
-    return !!this.userId
+  async _openDb () {
+    try {
+      this.userAppDb = common.utils.createPromisifiedHyperDB(this.userAppDbId)
+      await common.utils.dbReady(this.userAppDb)
+    } catch (e) {
+      if (this.isLoggedIn()) {
+        throw e
+      }
+    }
   }
 
-  isConnected () {
-    // is connected if the userAppDb has been created and initialized
-    return !!this.userAppDb
+  _createSwarm (hub) {
+    return swarm(hub, {
+      config: this.config.swarmConfig
+    })
   }
 
-  async connectToMasq () {
-    await this._loadSessionInfo()
-
-    if (!this.isLoggedIn()) {
-      await this._disconnect()
+  startReplication () {
+    debug('[masq.startReplication]')
+    if (this.state !== 'logged') {
       return
     }
 
     try {
-      const db = common.utils.createPromisifiedHyperDB(this.userId)
-      this.userAppDb = db
-      await common.utils.dbReady(db)
-      this._startReplication()
+      this.userAppDb.discoveryKey.toString('hex')
     } catch (e) {
-      if (this.isLoggedIn()) {
-        throw e
-      }
+      console.error(e)
     }
+    const discoveryKey = this.userAppDb.discoveryKey.toString('hex')
+    this.userAppRepHub = signalhub(discoveryKey, this.config.hubUrls)
+    this.userAppRepHub.on('error', () => {
+      debug('[masq.startReplication] dispatched event: replicationError')
+      this.eventTarget.dispatchEvent(new CustomEvent('replicationError', { detail: new MasqError(MasqError.REPLICATION_SIGNALLING_ERROR) }))
+    })
+    this.userAppRepSW = this._createSwarm(this.userAppRepHub)
+
+    this.setState('replicating')
+    this.userAppRepSW.on('peer', (peer, id) => {
+      try {
+        const stream = this.userAppDb.replicate({ live: true })
+        pump(peer, stream, peer)
+      } catch (e) {
+        if (this.isLoggedIn()) {
+          throw e
+        }
+      }
+    })
   }
 
-  // function useful to simulate the shutdown of a User-app
-  async _disconnect () {
+  async stopReplication () {
+    debug('[masq.stopReplication]')
     await new Promise((resolve) => {
-      this.userAppDb = null
       if (!this.userAppRepSW) {
```

Comment:
this check (`if (!this.userAppRepSW`) can be moved out from the promise

Reply:
:+1: